File: apps/services/notifications/Notifications.Infrastructure/Webhooks/Verifiers/TwilioVerifier.cs
using System.Security.Cryptography;
using System.Text;
using Microsoft.Extensions.Logging;

namespace Notifications.Infrastructure.Webhooks.Verifiers;

public class TwilioVerifier
{
    private readonly bool _enabled;
    private readonly string _authToken;
    private readonly string _environment;
    private readonly ILogger<TwilioVerifier> _logger;

    public TwilioVerifier(bool enabled, string authToken, string environment, ILogger<TwilioVerifier> logger)
    {
        _enabled = enabled;
        _authToken = authToken;
        _environment = environment;
        _logger = logger;
    }

    public (bool Verified, bool Skipped, string? Reason) Verify(string requestUrl, Dictionary<string, string> formParams, string? signature)
    {
        if (!_enabled)
        {
            _logger.LogWarning(
                "Twilio webhook verification is disabled; request rejected. " +
                "Set TWILIO_WEBHOOK_VERIFICATION_ENABLED=true and configure " +
                "TWILIO_AUTH_TOKEN to accept webhook callbacks.");
            return (false, false, "verification_disabled");
        }

        if (string.IsNullOrEmpty(_authToken))
        {
            _logger.LogError("Twilio webhook verification is enabled but no auth token is configured");
            return (false, false, "missing_auth_token");
        }

        if (string.IsNullOrEmpty(signature))
        {
            _logger.LogWarning("Twilio webhook request missing X-Twilio-Signature header");
            return (false, false, "missing_signature");
        }

        try
        {
            var sortedParams = string.Join("", formParams.OrderBy(kv => kv.Key).Select(kv => kv.Key + kv.Value));
            var dataToSign = requestUrl + sortedParams;

            using var hmac = new HMACSHA1(Encoding.UTF8.GetBytes(_authToken));
            var expectedSig = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(dataToSign)));

            var valid = CryptographicOperations.FixedTimeEquals(
                Convert.FromBase64String(signature),
                Convert.FromBase64String(expectedSig));

            if (!valid)
            {
                _logger.LogWarning("Twilio webhook signature verification failed");
                return (false, false, "invalid_signature");
            }

            return (true, false, null);
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Twilio webhook verification threw an error");
            return (false, false, $"verification_error: {ex.Message}");
        }
    }
}

File: _archived/notifications-nodejs/src/integrations/webhooks/verifiers/twilio.verifier.ts
import * as crypto from "crypto";
import { logger } from "../../../shared/logger";

interface TwilioVerifierConfig {
  enabled: boolean;
  authToken: string;
  nodeEnv: string;
}

export interface VerificationResult {
  verified: boolean;
  skipped: boolean;
  reason?: string;
}

/**
 * Verifies Twilio request signatures using HMAC-SHA1.
 *
 * Algorithm:
 *   1. Take the full URL of the request
 *   2. If POST: append sorted key=value pairs of the POST parameters
 *   3. Sign with HMAC-SHA1 using the Twilio auth token
 *   4. Compare base64 result with X-Twilio-Signature header
 */
export class TwilioWebhookVerifier {
  private config: TwilioVerifierConfig;

  constructor(config: TwilioVerifierConfig) {
    this.config = config;
  }

  verify(
    requestUrl: string,
    params: Record<string, string>,
    signature: string | undefined
  ): VerificationResult {
    const { enabled, authToken, nodeEnv } = this.config;

    if (!enabled) {
      if (nodeEnv === "production") {
        logger.warn(
          "Twilio webhook verification is disabled in production — requests are accepted without signature check"
        );
      }
      return { verified: false, skipped: true, reason: "verification_disabled" };
    }

    if (!authToken) {
      logger.error("Twilio webhook verification is enabled but no auth token is configured");
      return { verified: false, skipped: false, reason: "missing_auth_token" };
    }

    if (!signature) {
      logger.warn("Twilio webhook request missing X-Twilio-Signature header");
      return { verified: false, skipped: false, reason: "missing_signature" };
    }

    try {
      // Build the string to sign: URL + sorted params concatenated
      const sortedParams = Object.keys(params)
        .sort()
        .reduce((str, key) => str + key + (params[key] ?? ""), requestUrl);

      const expectedSig = crypto
        .createHmac("sha1", authToken)
        .update(sortedParams, "utf8")
        .digest("base64");

      const valid = crypto.timingSafeEqual(
        Buffer.from(signature, "base64"),
        Buffer.from(expectedSig, "base64")
      );

      if (!valid) {
        logger.warn("Twilio webhook signature verification failed");
        return { verified: false, skipped: false, reason: "invalid_signature" };
      }

      return { verified: true, skipped: false };
    } catch (err) {
      logger.error("Twilio webhook verification threw an error", { error: String(err) });
      return { verified: false, skipped: false, reason: `verification_error: ${String(err)}` };
    }
  }
}
